include/deal.II/matrix_free/operators.h

```diff
@@ -205,6 +372,299 @@ namespace MatrixFreeOperators
+    mg_constrained_dofs.get_refinement_edge_indices(level).fill_index_vector(interface_indices);
+    edge_constrained_indices.clear();
+    edge_constrained_indices.reserve(interface_indices.size());
+    edge_constrained_values.resize(interface_indices.size());
+    const IndexSet &locally_owned = data->get_dof_handler().locally_owned_mg_dofs(level);
+    for (unsigned int i=0; i<interface_indices.size(); ++i)
+      if (locally_owned.is_element(interface_indices[i]))
+        edge_constrained_indices.push_back(locally_owned.index_within_set(interface_indices[i]));
+    have_interface_matrices = Utilities::MPI::max((unsigned int)edge_constrained_indices.size(),
+                                                  data_.get_vector_partitioner()->get_communicator()) > 0;
+  }
+
+  template <int dim, typename number>
+  void
+  Base<dim,number>::vmult (LinearAlgebra::distributed::Vector<number>       &dst,
+                           const LinearAlgebra::distributed::Vector<number> &src) const
+  {
+    dst = 0;
+    vmult_add (dst, src);
+  }
+
+  template <int dim, typename number>
+  void
+  Base<dim,number>::vmult_add (LinearAlgebra::distributed::Vector<number> &dst,
+                               const LinearAlgebra::distributed::Vector<number> &src) const
+  {
+    Assert(src.partitioners_are_globally_compatible(*data->get_dof_info(0).vector_partitioner), ExcInternalError());
+    Assert(dst.partitioners_are_globally_compatible(*data->get_dof_info(0).vector_partitioner), ExcInternalError());
+
+    // set zero Dirichlet values on the input vector (and remember the src and
+    // dst values because we need to reset them at the end)
+    for (unsigned int i=0; i<edge_constrained_indices.size(); ++i)
+      {
+        edge_constrained_values[i] =
+          std::pair<number,number>(src.local_element(edge_constrained_indices[i]),
+                                   dst.local_element(edge_constrained_indices[i]));
+        const_cast<LinearAlgebra::distributed::Vector<number>&>(src).local_element(edge_constrained_indices[i]) = 0.;
+      }
+
+    apply_add(dst,src);
+
+    const std::vector<unsigned int> &
+    constrained_dofs = data->get_constrained_dofs();
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      dst.local_element(constrained_dofs[i]) += src.local_element(constrained_dofs[i]);
+
+    // reset edge constrained values, multiply by unit matrix and add into
+    // destination
+    for (unsigned int i=0; i<edge_constrained_indices.size(); ++i)
+      {
+        const_cast<LinearAlgebra::distributed::Vector<number>&>(src).local_element(edge_constrained_indices[i]) = edge_constrained_values[i].first;
+        dst.local_element(edge_constrained_indices[i]) = edge_constrained_values[i].second + edge_constrained_values[i].first;
+      }
+  }
+
+
+  template <int dim, typename number>
+  void
+  Base<dim,number>::Tvmult_add (LinearAlgebra::distributed::Vector<number> &dst,
```

Comment:
You use this because you would otherwise have to duplicate the manipulations in the vectors and the constraints, right?